test/libsolidity/util/TestFileParser.cpp

```diff
@@ -0,0 +1,437 @@
+		/// its string representation still contains a sinlge `)`, because
+		/// that's the end of recursion in `parseIdentifierOrTuple`.
+		/// If already appended to the parameters, do not append again.
+		const bool isListEmpty = parameters == formatToken(SoltToken::RParen);
+		if (!isListEmpty)
+			signature += formatToken(SoltToken::RParen);
+	}
+	return signature;
+}
+
+u256 TestFileParser::parseFunctionCallValue()
+{
+	u256 value = convertNumber(parseNumber());
+	expect(SoltToken::Ether);
+	return value;
+}
+
+FunctionCallArgs TestFileParser::parseFunctionCallArguments()
+{
+	FunctionCallArgs arguments;
+
+	auto param = parseParameter();
+	if (param.abiType.type == ABIType::None)
+		throw Error(Error::Type::ParserError, "No argument provided.");
+	arguments.parameters.emplace_back(param);
+
+	while (accept(SoltToken::Comma, true))
+		arguments.parameters.emplace_back(parseParameter());
+	return arguments;
+}
+
+FunctionCallExpectations TestFileParser::parseFunctionCallExpectations()
+{
+	FunctionCallExpectations expectations;
+
+	auto param = parseParameter();
+	if (param.abiType.type == ABIType::None)
+	{
+		expectations.failure = false;
+		return expectations;
+	}
+	expectations.result.emplace_back(param);
+
+	while (accept(SoltToken::Comma, true))
+		expectations.result.emplace_back(parseParameter());
+
+	/// We have always one virtual parameter in the parameter list.
+	/// If its type is FAILURE, the expected result is also a REVERT etc.
+	if (expectations.result.at(0).abiType.type != ABIType::Failure)
+		expectations.failure = false;
+	return expectations;
+}
+
+Parameter TestFileParser::parseParameter()
+{
+	Parameter parameter;
+	if (accept(SoltToken::Newline, true))
+		parameter.format.newline = true;
+	auto literal = parseABITypeLiteral();
+	parameter.rawBytes = literal.first;
+	parameter.abiType = literal.second;
+	return parameter;
+}
+
+pair<bytes, ABIType> TestFileParser::parseABITypeLiteral()
+{
+	try
+	{
+		u256 number{0};
+		ABIType abiType{ABIType::None, 0};
+
+		if (accept(SoltToken::Sub))
+		{
+			abiType = ABIType{ABIType::SignedDec, 32};
+			expect(SoltToken::Sub);
+			number = convertNumber(parseNumber()) * -1;
+		}
+		else
+		{
+			if (accept(SoltToken::Number))
+			{
+				abiType = ABIType{ABIType::UnsignedDec, 32};
+				number = convertNumber(parseNumber());
+			}
+			else if (accept(SoltToken::Failure, true))
+			{
+				abiType = ABIType{ABIType::Failure, 0};
+				return make_pair(bytes{}, abiType);
+			}
+		}
+		return make_pair(toBigEndian(number), abiType);
+	}
+	catch (std::exception const&)
+	{
+		throw Error(Error::Type::ParserError, "Number encoding invalid.");
+	}
+}
+
+string TestFileParser::parseIdentifierOrTuple()
+{
+	string identOrTuple;
+	if (accept(SoltToken::Identifier))
+	{
+		identOrTuple = m_scanner.currentLiteral();
+		expect(SoltToken::Identifier);
+		return identOrTuple;
+	}
+	/// This is the end of recursion.
+	if (accept(SoltToken::RParen, false))
```

Comment:
Does/Can this in fact happen in valid cases?

Reply:
At least if you don't have somthing like `f(: 1 ->`. And we need to cover `f((): 1 ->`.

Reply:
I'm not sure that I understand your thought there... sorry for not just letting it be ;-), but:

My point is: the identifier case is already done above, isn't it? Why can't you just ``expect(SoltToken::LParen);`` here? Then you could skip the ``isTuple`` flag and the ``if (accept(SoltToken::LParen, true))`` below and replace it by unconditionally doing ``identOrTuple += formatToken(SoltToken::LParen);``. Then line 281 to 290 can just stay the same and 292-296 can just be replaced by ``expect(SoltToken::RParen);``.

Reply:
Thanks! Fixed.